### PR TITLE
Config check before any set requests issued

### PR DESCRIPTION
**Pull request by Andrea-Campanella**
Checking configuration validity against models before applying any to any device. Fixes:#543

### CODE REVIEW

File: pkg/northbound/gnmi/set.go [hunk 1]
```diff
@@ -98,6 +98,23 @@ func (s *Server) Set(ctx context.Context, req *gnmi.SetRequest) (*gnmi.SetRespon
 		return nil, status.Error(codes.InvalidArgument, errRo.Error())
 	}
 
+	//Checking for wrong configuration against the device models for updates
+	for target, updates := range targetUpdates {
+		_, err := validateChange(target, version, updates, targetRemoves[target])
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set.go]: you could parallelize this like I do in manager.go ValidateStores()

File: pkg/northbound/gnmi/set.go [hunk 2]
```diff
@@ -98,6 +98,23 @@ func (s *Server) Set(ctx context.Context, req *gnmi.SetRequest) (*gnmi.SetRespon
...
+		if err != nil {
+			return nil, err
+		}
+		//TODO copy targetRemoves into a temporary set and delete from that and use it at line 112
+		//delete(targetRemoves, target)
+	}
+	//Checking for wrong configuration against the device models for deletes
+	for target, removes := range targetRemoves {
+		_, err := validateChange(target, version, make(map[string]*change.TypedValue), removes)
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set.go]: same as above

File: pkg/manager/setconfig.go [hunk 3]
```diff
@@ -115,3 +110,40 @@ func (m *Manager) SetNetworkConfig(configName store.ConfigName, updates map[stri
 
 	return changeID, configName, nil
 }
+
+func (m *Manager) getStoredConfig(configName store.ConfigName) (store.Configuration, store.ConfigName, error) {
+	// Look for an exact match or then a partial match
```
Comment (SeanCondon) [inline pkg/manager/setconfig.go]: maybe put the comment before the func

File: pkg/manager/setconfig.go [hunk 4]
```diff
@@ -27,85 +27,80 @@ import (
...
-		for key := range m.ConfigStore.Store {
-			deviceNameFromKey := string(key)[:strings.LastIndex(string(key), "-")]
-			if deviceNameFromKey == string(configName) {
-				similarDevices = append(similarDevices, key)
-			}
-		}
+// ValidateNetworkConfig validates the given updates and deletes, according to the path on the configuration
+// for the specified target
+func (m *Manager) ValidateNetworkConfig(configName store.ConfigName, updates map[string]*change.TypedValue,
+	deletes []string) (change.ID, error) {
```
Comment (SeanCondon) [inline pkg/manager/setconfig.go]: now that you don't store it I think there's no point at all in returning this change.ID

File: pkg/northbound/gnmi/set.go [hunk 5]
```diff
@@ -386,3 +403,18 @@ func setChange(target string, version string, targetUpdates map[string]*change.T
 	}
 	return changeID, configName, false, nil
 }
+
+func validateChange(target string, version string, targetUpdates map[string]*change.TypedValue, targetRemoves []string) (change.ID, error) {
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set.go]: again, because you don't store the change I think there's no point in returning it

Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set.go]: done

File: pkg/manager/setconfig.go (hunk 4 above)
Comment (Andrea-Campanella) [reply pkg/manager/setconfig.go]: done

**Pull request merged**